### PR TITLE
[SPARK-42519][CONNECT][TESTS] Add More WriteTo Tests In Spark Connect Client

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
Add more WriteTo tests for Spark Connect Client

### Why are the changes needed?
Improve Test Case, remove same todo

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new tests

### CODE REVIEW

**Comment by amaliujia** [comment]
cc @zhenlineo can you take a look?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala [hunk 1]
```diff
@@ -58,10 +58,12 @@ object SparkConnectServerUtils {
 
   private lazy val sparkConnect: Process = {
     debug("Starting the Spark Connect Server...")
-    val jar = findJar(
+    val connectJar = findJar(
       "connector/connect/server",
       "spark-connect-assembly",
       "spark-connect").getCanonicalPath
+    val catalystTestJar =
+      findJar("sql/catalyst", "spark-catalyst", "spark-catalyst", test = true).getCanonicalPath
```
Comment (zhenlineo) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:66]: This would not work with a SBT only build, right?
Does this mean, for all SBT users, they have to build with maven once? Can you make sure the following case works:
```
build/sbt clean
build/sbt "testOnly org.apache.spark.sql.ClientE2ETestSuite”
```

Can you point me which is the test jar built by SBT?

**Review by zhenlineo — COMMENTED** [review]
Thanks so much for the more test cases added. The change looks great. I only have a question about how to find the test jar for the SBT build.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 2]
```diff
@@ -683,7 +770,7 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper {
     checkSameResult(list.asScala, session.createDataset(list))
     checkSameResult(
       list.asScala.map(kv => Row(kv.key, kv.value)),
-      session.createDataFrame(list.asScala.toSeq))
+      session.createDataFrame(list.asScala))
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala]: Thanks. Done

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala [hunk 3]
```diff
@@ -65,29 +65,38 @@ object IntegrationTestUtils {
...
-        f.getName.endsWith(s"${org.apache.spark.SPARK_VERSION}.jar"))
+        f.getName.endsWith(s"${org.apache.spark.SPARK_VERSION}$suffix"))
     }
     // It is possible we found more than one: one built by maven, and another by SBT
     assert(jars.nonEmpty, s"Failed to find the jar inside folder: ${targetDir.getCanonicalPath}")
     debug("Using jar: " + jars(0).getCanonicalPath)
     jars(0) // return the first jar found
   }
 
+  private[sql] def findJar(path: String, sbtName: String, mvnName: String): File = {
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala]: Thanks. Done

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala (hunk 1 above)
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:66]: I try build with SBT locally. But I can't find where is test jar in target/scala-2.13. Should I change SBT build logic to add `sbt test:package`?

Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala]: good advise. done

Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala]: good advise. done

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala [hunk 4]
```diff
@@ -57,6 +57,12 @@ object IntegrationTestUtils {
     Files.exists(Paths.get(filePath))
   }
 
+  private[sql] lazy val isCatalystTestJarAvailable: Boolean = {
+    val filePath = s"$sparkHome/sql/catalyst/target/" +
+      s"spark-catalyst_$scalaVersion-${org.apache.spark.SPARK_VERSION}-tests.jar"
```
[the remark this answers is not in the document]
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala]: Thanks for help! I changed the code.

**Review by zhenlineo — APPROVED** [review]
LGTM.

**Comment by zhenlineo** [comment]
cc @LuciferYang Can we merge this?

**Comment by LuciferYang** [comment]
The pr title should be `[SPARK-42519][CONNECT][TESTS] ...`

**Comment by LuciferYang** [comment]
@zhenlineo Thanks for ping me ~ Let me do some manual checks

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala [hunk 5]
```diff
@@ -434,7 +433,9 @@ abstract class InMemoryBaseTable(
     protected var streamingWriter: StreamingWrite = StreamingAppend
 
     override def overwriteDynamicPartitions(): WriteBuilder = {
-      assert(writer == Append)
+      if (writer != Append) {
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:436]: What is the reason for changing the exception type?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala [hunk 6]
```diff
@@ -58,10 +58,15 @@ object SparkConnectServerUtils {
 
   private lazy val sparkConnect: Process = {
     debug("Starting the Spark Connect Server...")
-    val jar = findJar(
+    val connectJar = findJar(
       "connector/connect/server",
       "spark-connect-assembly",
       "spark-connect").getCanonicalPath
+    var driverClassPath = connectJar
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:65]: I think we can make `driverClassPath` as `val` like:
```
val driverClassPath = f (IntegrationTestUtils.isCatalystTestJarAvailable) {
    connectJar += ":" ...
} else connectJar
```

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 7]
```diff
@@ -220,41 +220,132 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper {
     }
   }
 
+  test("writeTo with create") {
+    assume(IntegrationTestUtils.isCatalystTestJarAvailable)
+    withTable("testcat.myTableV2") {
+      spark.conf.set(
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:226]: can we use `withSQLConf` ?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 8]
```diff
@@ -220,41 +220,132 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper {
     }
   }
 
+  test("writeTo with create") {
+    assume(IntegrationTestUtils.isCatalystTestJarAvailable)
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:224]: local run `build/sbt "connect-client-jvm/testOnly *ClientE2ETestSuite" -Phive`, the new test always be `IGNORED`.

Do you mind refactor the following code to make local sbt test can test them as default? 

https://github.com/apache/spark/blob/74cddcfda3ac4779de80696cdae2ba64d53fc635/project/SparkBuild.scala#L855-L857

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 7 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:226]: Or should we make this the default config for `SimpleSparkConnectService` startup, then we can skip this manual configuration

@zhenlineo WDYT?

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 8 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:224]: For example , add `(LocalProject("catalyst") / Test / Keys.`package`).value` ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala (hunk 5 above)
Comment (Hisoka-X) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:436]: The reason is we want to use catalyst test jar to submit ConnectService, so we can use `InMemoryTableCatalog`. But if `scalatest` used by `InMemoryTableCatalog`. We should add `scalatest` jar into ConnectService too. So I remove it.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala (hunk 6 above)
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala]: Thanks for advise. Finished

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 7 above)
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala]: I removed it and change `RemoteSparkSession` to set config `spark.sql.catalog.testcat` as default.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala (hunk 8 above)
Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:224]: Done.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala [hunk 9]
```diff
@@ -57,6 +57,15 @@ object IntegrationTestUtils {
     Files.exists(Paths.get(filePath))
   }
 
+  private[sql] lazy val isCatalystTestJarAvailable: Boolean = {
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala:60]: I want to know which scenarios `isCatalystTestJarAvailable` may be false now

Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala:60]: Build without test jar?

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala:60]: I test the following scenarios:

1. build/sbt clean "connect-client-jvm/test"         
2. build/sbt clean "connect-client-jvm/testOnly *ClientE2ETestSuite"

3. 
```
build/mvn clean
build/mvn clean package test -Dtest=none -DwildcardSuites=org.apache.spark.sql.ClientE2ETestSuite
```
4.
```
build/mvn clean
build/mvn clean install -pl connector/connect/server -am -DskipTests   
build/mvn clean install -pl assembly -DskipTests
build/mvn clean test -pl connector/connect/client/jvm
``` 

New tests can be triggered and tested successfully. 

@Hisoka-X So haver you thought of any other way to successfully test before this PR, but without `isCatalystTestJarAvailable ` checking, the test will fail?

Comment (Hisoka-X) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala:60]: Before add `isCatalystTestJarAvailable` check, I didn't modify sbt build script. So it didn't work on sbt . Check https://github.com/apache/spark/pull/40564#discussion_r1152081937 . But now I think I can remove this check.

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 10]
```diff
@@ -670,6 +738,7 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper {
       bean.setValue(v)
       bean
     }
+
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:741]: This seems unrelated to the current pr, please revert this blank line

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 11]
```diff
@@ -801,7 +870,9 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper {
 }
 
 private[sql] case class MyType(id: Long, a: Double, b: Double)
+
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:873]: ditto

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 12]
```diff
@@ -801,7 +870,9 @@ class ClientE2ETestSuite extends RemoteSparkSession with SQLHelper {
...
 private[sql] case class KV(key: String, value: Int)
+
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:875]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala (hunk 5 above)
Comment (LuciferYang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:436]: make sense

**Review by LuciferYang — APPROVED** [review]
+1, LGTM except for some new comments

**Comment by LuciferYang** [comment]
```
### How was this patch tested?
Yes
```

I think You can say `Add new tests`

**Comment by LuciferYang** [comment]
cc @hvanhovell @HyukjinKwon @zhengruifeng FYI

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Comment by Hisoka-X** [comment]
Thanks @zhenlineo @LuciferYang @HyukjinKwon
